File: objc/Project11/Project11/Bacteria.m
//
//  Bacteria.m
//  Project11
//
//  Created by Jinwoo Kim on 2/20/21.
//

#import "Bacteria.h"

@interface Bacteria ()
@property UIView *connection;
@end

@implementation Bacteria
@synthesize color = _color;
@synthesize direction = _direction;

+(instancetype)buttonWithType:(UIButtonType)buttonType {
    Bacteria *result = [super buttonWithType:buttonType];
    
    if (result) {
        result.connection = [[UIView alloc] initWithFrame:CGRectMake(27, -11, 16, 12)];
        result.row = 0;
        result.col = 0;
    }
    
    return result;
}

//- (instancetype)init {
//    self = [self init];
//
//    if (self) {
//        self.connection = [[UIView alloc] initWithFrame:CGRectMake(27, -11, 16, 12)];
//        self.row = 0;
//        self.col = 0;
//        self.color = 0;
//        _color = UIColor.grayColor;
//        _direction = DirectionTypeNorth;
//    }
//
//    return self;
//}

- (UIColor *)color {
    // reading: just return the private value
    return _color;
}

- (void)setColor:(UIColor *)color {
    // writing: update the private value
    _color = color;
    
    // make the connection value change its color
    self.connection.backgroundColor = color;
    
    // figure out whice image to use and apply it immediately
    if (color == UIColor.greenColor) {
        [self setImage:[UIImage imageNamed:@"arrowGreen"] forState:UIControlStateNormal];
    } else if (color == UIColor.redColor) {
        [self setImage:[UIImage imageNamed:@"arrowRed"] forState:UIControlStateNormal];
    } else {
        [self setImage:[UIImage imageNamed:@"arrowGray"] forState:UIControlStateNormal];
    }
}

- (DirectionType)direction {
    // reading: just return the private value
    return _direction;
}

- (void)setDirection:(DirectionType)direction {
    // writing: update the private value
    _direction = direction;
    
    // make ourselves point in the correct direction
    switch (direction) {
        case DirectionTypeNorth:
            self.transform = CGAffineTransformIdentity;
            break;
        case DirectionTypeEast:
            self.transform = CGAffineTransformMakeRotation(M_PI / 2);
            break;
        case DirectionTypeSouth:
            self.transform = CGAffineTransformMakeRotation(M_PI);
            break;
        case DirectionTypeWest:
            self.transform = CGAffineTransformMakeRotation(M_PI + M_PI / 2);
            break;
        default:
            break;
    }
}

- (void)addConnection {
    [self addSubview:self.connection];
}

- (void)rotate {
    switch (self.direction) {
        case DirectionTypeNorth:
            self.direction = DirectionTypeEast;
            break;
        case DirectionTypeEast:
            self.direction = DirectionTypeSouth;
            break;
        case DirectionTypeSouth:
            self.direction = DirectionTypeWest;
            break;
        case DirectionTypeWest:
            self.direction = DirectionTypeNorth;
            break;
        default:
            break;
    }
}

@end

File: swift/Project11/Project11/Bacteria.swift
//
//  Bacteria.swift
//  Project11
//
//  Created by Jinwoo Kim on 2/18/21.
//

import UIKit

enum Direction {
    case north, south, east, west
}

class Bacteria: UIButton {
    let connection = UIView(frame: CGRect(x: 27, y: -11, width: 16, height: 12))
    var row = 0
    var col = 0
    
    private var _color = UIColor.gray
    var color: UIColor {
        get {
            // reading: just return the private value
            return _color
        }
        
        set {
            // writing: update the private value
            _color = newValue
            
            // make the conntection value change its color
            connection.backgroundColor = newValue
            
            // figure out which image to use and apply it immediately
            switch newValue {
            case UIColor.green:
                setImage(UIImage(named: "arrowGreen"), for: .normal)
            case UIColor.red:
                setImage(UIImage(named: "arrowRed"), for: .normal)
            default:
                setImage(UIImage(named: "arrowGray"), for: .normal)
            }
        }
    }
    
    private var _direction = Direction.north
    var direction: Direction {
        get {
            // reading: just return the private value
            return _direction
        }
        
        set {
            // writing: update the private value
            _direction = newValue
            
            // make ourselves point in the correct direction
            switch newValue {
            case .north:
                transform = .identity
            case .east:
                transform = CGAffineTransform(rotationAngle: CGFloat.pi / 2)
            case .south:
                transform = CGAffineTransform(rotationAngle: CGFloat.pi)
            case .west:
                transform = CGAffineTransform(rotationAngle: CGFloat.pi + CGFloat.pi / 2)
            }
        }
    }
    
    func addConnection() {
        addSubview(connection)
    }
    
    func rotate() {
        switch direction {
        case .north:
            direction = .east
        case .east:
            direction = .south
        case .south:
            direction = .west
        case .west:
            direction = .north
        }
    }
}
